cli/azd/internal/telemetry/storage.go

```diff
@@ -0,0 +1,279 @@
+	randomSuffix := generatedFileName[:strings.LastIndex(generatedFileName, "_")]
+	readyTime := stg.clock.Now().Add(delayDuration)
+	fileName := formatFileName(readyTime, retryCount, randomSuffix, stg.itemFileExtension)
+	err = os.Rename(file.Name(), filepath.Join(stg.folder, fileName))
+	if err != nil {
+		return fmt.Errorf("failed to rename file: %v", err)
+	}
+
+	return nil
+}
+
+// Gets the next available item for processing.
+// Returns nil if no items exist.
+// Returns error if an error occurs while reading storage.
+func (stg *StorageQueue) Peek() (*StoredItem, error) {
+	items, err := stg.getAllItems()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get stored files: %v", err)
+	}
+
+	leastRecentTime := time.Time{}
+	latestIndex := -1
+	now := stg.clock.Now()
+	for i, item := range items {
+		if now.Sub(item.readyTime) >= 0 {
+			if latestIndex == -1 || item.fileModTime.Before(leastRecentTime) {
+				leastRecentTime = item.fileModTime
+				latestIndex = i
+			}
+		}
+	}
+
+	if latestIndex == -1 {
+		return nil, nil
+	}
+
+	item := items[latestIndex]
+	fileName := filepath.Join(stg.folder, item.name)
+	message, err := os.ReadFile(fileName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read latest stored item: %v", err)
+	}
+
+	return &StoredItem{
+		fileName:   fileName,
+		retryCount: item.retryCount,
+		message:    message,
+	}, nil
+}
+
+// Removes the stored item from queue.
+// Does not return an error if the item is already removed.
+func (stg *StorageQueue) Remove(item *StoredItem) error {
+	err := os.Remove(item.fileName)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to remove stored item: %v", err)
+	}
+
+	return nil
+}
+
+// Scans the storage directory for any obsoleted items or temp files.
+func (stg *StorageQueue) Cleanup() {
+	files, err := os.ReadDir(stg.folder)
+	if err != nil {
+		return
+	}
+
+	for _, file := range files {
+		if !file.IsDir() {
+			name := file.Name()
+			if strings.HasSuffix(name, ".tmp") {
+				info, err := file.Info()
+				if err == nil && stg.clock.Since(info.ModTime()) > tempFileTtl {
+					_ = os.Remove(filepath.Join(stg.folder, name))
+				}
+			} else if strings.HasSuffix(name, stg.itemFileExtension) {
+				if _, ok := parseFileName(name); !ok {
+					_ = os.Remove(filepath.Join(stg.folder, name))
+				}
+			}
+		}
+	}
+}
+
+func (stg *StorageQueue) getAllItems() ([]itemEntry, error) {
+	dirEntries, err := readDir(stg.folder)
+	if err != nil {
+		return nil, fmt.Errorf("error reading folder: %v", err)
+	}
+
+	items := []itemEntry{}
+
+	for _, entry := range dirEntries {
+		if !entry.IsDir() && strings.HasSuffix(entry.Name(), stg.itemFileExtension) {
+			item, ok := readItemEntry(entry)
+
+			if ok {
+				items = append(items, *item)
+			}
+		}
+	}
+
+	return items, nil
+}
+
+// readDir is os.ReadDir except it returns entries in directory order instead of by name order
+func readDir(name string) ([]os.DirEntry, error) {
```

Comment:
Is there a reason we need this ordering instead of name order as `os.ReadDir` would give us?

Reply:
Just for performance, since we end up sorting by `modTime` instead.

Reply:
Gotcha, slight preference for just removing this in favor of `os.ReadDir`.  My concern is basically that the order returned here is unspecified (I suspect in cases where you'd added and removed files over time the ordering will end up not being by `modTime` but instead will just be in whatever order the filesystem has them, which may differ.

Reply:
The order here is definitely unspecified. It's later ordered by `modTime` when needed in `Peek()`. This is done to save the sort in lieu of a single O(n) loop comparison.

I think we have two options for better maintainability:
1. Add sorting here for by `fileModTime` so that this is consistently ordered.
2. Rename the methods here to `readDirUnordered` and `getAllItemsUnordered`, and keep the current implementation.

I think 2. makes more sense currently and will do that.